syntax = "proto3";

import "google/protobuf/struct.proto";

package fraud_Detection;

service FraudDetectionService {
    // New method to store or initialize the order data
    rpc InitializeOrder (FraudInitRequest) returns (FraudInitResponse);

    // Existing method that does the real fraud check
    rpc CheckFraud (FraudRequest) returns (FraudResponse);

    rpc ClearOrder(ClearOrderRequest) returns (ClearOrderResponse);
}

// New message for the initialization phase
message FraudInitRequest {
    double amount = 1;
    string ip_address = 2;
    string email = 3;
    string billing_country = 4;
    string billing_city = 5;
    string payment_method = 6;
    string order_id = 7;
    map<string, int32> vectorClock = 8; 
}

message FraudInitResponse {
    bool success = 1;
    map<string, int32> vectorClock = 2; 
}


// Existing
message FraudRequest {
    double amount = 1;
    string ip_address = 2;
    string email = 3;
    string billing_country = 4;
    string billing_city = 5;
    string payment_method = 6;
    string order_id = 7;
    map<string, int32> vectorClock = 8; 
}

message FraudResponse {
    double fraud_probability = 1;
    string action = 2;
    google.protobuf.Struct details = 3;
    repeated string reasons = 4;
    map<string, int32> vectorClock = 5; 
}

message ClearOrderRequest {
    string order_id = 1;
    map<string, int32> vectorClock = 2;
}

message ClearOrderResponse {
    bool success = 1;
    string error = 2;
    map<string, int32> vectorClock = 3;
}
